rfd/0229a-scoped-join-tokens.md

```diff
@@ -0,0 +1,331 @@
+- Support for configuring a scoped token with a set of labels that are
+  automatically assigned to SSH nodes at provisioning time.
+- Support for single use (oneshot) tokens.
+- New `scoped` variants of the `tctl tokens *` family of sub commands.
+
+Considered out of scope for this RFD:
+
+- Token types other than static and ephemeral tokens or join methods other than
+  "token". Other join methods will almost certainly be implemented in the
+  future but are not part of the immediately planned work.
+
+### Scoped tokens
+
+The recently added `ScopedToken` resource will be modified to include fields
+necessary for assigning scopes, labels, and enforcing single use. Additionally,
+fields missing from the existing `types.ProvisionTokenV2` will be ported over
+to facilitate existing provisioning semantics.
+
+```diff
+--- a/api/proto/teleport/scopes/joining/v1/token.proto
++++ b/api/proto/teleport/scopes/joining/v1/token.proto
+@@ -16,6 +16,7 @@ syntax = "proto3";
+
+ package teleport.scopes.joining.v1;
+
++import "google/protobuf/timestamp.proto";
+ import "teleport/header/v1/metadata.proto";
+
+ option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/scopes/joining/v1;joiningv1";
+@@ -42,12 +43,48 @@ message ScopedToken {
+
+   // Spec is the token specification.
+   ScopedTokenSpec spec = 6;
++
++  // The status of the token.
++  ScopedTokenStatus status = 7;
+ }
+
+ // ScopedTokenSpec is the specification of a scoped token.
+ message ScopedTokenSpec {
+-  // AssignedScope is the scope to which this token is assigned.
++  // The scope to which this token is assigned.
+   string assigned_scope = 1;
+
+-  // TODO(fspmarshall): port relevant token features to scoped tokens.
++  // The list of roles associated with the token. They will be converted
++  // to metadata in the SSH and X509 certificates issued to the user of the
++  // token.
++  repeated string roles = 2;
++
++  // The joining method required in order to use this token.
++  // Supported joining methods for scoped tokens only include 'token'.
++  string join_method = 3;
++
++  // The usage mode of the token. Can be "oneshot" or "unlimited". Oneshot
++  // tokens can only be used to provision a single resource. Unlimited tokens
++  // can be be used to provision any number of resources until it expires.
++  string mode = 4;
++
++  // Immutable labels that should be encoded into the certificate of an
++  // SSH node provisioned using this token.
++  map<string, string> ssh_labels = 5;
++}
++
++// The usage status of a oneshot token.
++message OneshotStatus {
++  // The timestamp representing when a oneshot token was successfully used for
++  // provisioning.
++  google.protobuf.Timestamp used_at = 1;
++  // The timestamp representing when a oneshot token should no longer be available
++  // for idempotent retries.
++  google.protobuf.Timestamp reusable_until = 2;
++  // The public key of the host that used the token.
++  bytes used_by_public_key = 3;
```

Comment:
This can be a fingerprint of the public key, to save some space and to avoid having to think about the security of disclosing the public key in the token status.